File: lib/models/mega_evento.dart
import 'dart:convert';

class MegaEvento {
  final int megaEventoId;
  final String titulo;
  final String? descripcion;
  final DateTime fechaInicio;
  final DateTime fechaFin;
  final String? ubicacion;
  final double? lat;
  final double? lng;
  final String? categoria;
  final String estado;
  final int ongOrganizadoraPrincipal;
  final int? capacidadMaxima;
  final bool esPublico;
  final bool activo;
  final List<dynamic>? imagenes;
  final DateTime? fechaCreacion;
  final DateTime? fechaActualizacion;
  final Map<String, dynamic>? ongPrincipal;

  MegaEvento({
    required this.megaEventoId,
    required this.titulo,
    this.descripcion,
    required this.fechaInicio,
    required this.fechaFin,
    this.ubicacion,
    this.lat,
    this.lng,
    this.categoria,
    required this.estado,
    required this.ongOrganizadoraPrincipal,
    this.capacidadMaxima,
    required this.esPublico,
    required this.activo,
    this.imagenes,
    this.fechaCreacion,
    this.fechaActualizacion,
    this.ongPrincipal,
  });

  // Helper para parsear arrays que pueden venir como JSON string o array
  static List<dynamic>? _parseJsonArray(dynamic value) {
    if (value == null) return null;
    if (value is List) {
      final filtered =
          value
              .where(
                (item) =>
                    item != null &&
                    item.toString().trim().isNotEmpty &&
                    item.toString() != '[]' &&
                    item.toString() != 'null',
              )
              .toList();
      return filtered.isEmpty ? null : filtered;
    }
    if (value is String) {
      final trimmed = value.trim();
      if (trimmed.isEmpty || trimmed == '[]' || trimmed == 'null') {
        return null;
      }

      try {
        final decoded = jsonDecode(value);
        if (decoded is List) {
          final filtered =
              decoded
                  .where(
                    (item) =>
                        item != null &&
                        item.toString().trim().isNotEmpty &&
                        item.toString() != '[]' &&
                        item.toString() != 'null',
                  )
                  .toList();
          return filtered.isEmpty ? null : filtered;
        }
        return [value];
      } catch (e) {
        return [value];
      }
    }
    return [value];
  }

  factory MegaEvento.fromJson(Map<String, dynamic> json) {
    return MegaEvento(
      megaEventoId: json['mega_evento_id'] as int,
      titulo: json['titulo'] as String,
      descripcion: json['descripcion'] as String?,
      fechaInicio: DateTime.parse(json['fecha_inicio'] as String),
      fechaFin: DateTime.parse(json['fecha_fin'] as String),
      ubicacion: json['ubicacion'] as String?,
      lat:
          json['lat'] != null
              ? (json['lat'] is String
                  ? double.tryParse(json['lat'] as String)
                  : (json['lat'] as num).toDouble())
              : null,
      lng:
          json['lng'] != null
              ? (json['lng'] is String
                  ? double.tryParse(json['lng'] as String)
                  : (json['lng'] as num).toDouble())
              : null,
      categoria: json['categoria'] as String?,
      estado: json['estado'] as String,
      ongOrganizadoraPrincipal: json['ong_organizadora_principal'] as int,
      capacidadMaxima: json['capacidad_maxima'] as int?,
      esPublico: json['es_publico'] == 1 || json['es_publico'] == true,
      activo: json['activo'] == 1 || json['activo'] == true,
      imagenes: _parseJsonArray(json['imagenes']),
      fechaCreacion:
          json['fecha_creacion'] != null
              ? DateTime.parse(json['fecha_creacion'] as String)
              : null,
      fechaActualizacion:
          json['fecha_actualizacion'] != null
              ? DateTime.parse(json['fecha_actualizacion'] as String)
              : null,
      ongPrincipal:
          json['ong_principal'] != null
              ? json['ong_principal'] as Map<String, dynamic>
              : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'mega_evento_id': megaEventoId,
      'titulo': titulo,
      'descripcion': descripcion,
      'fecha_inicio': fechaInicio.toIso8601String(),
      'fecha_fin': fechaFin.toIso8601String(),
      'ubicacion': ubicacion,
      'lat': lat,
      'lng': lng,
      'categoria': categoria,
      'estado': estado,
      'ong_organizadora_principal': ongOrganizadoraPrincipal,
      'capacidad_maxima': capacidadMaxima,
      'es_publico': esPublico,
      'activo': activo,
      'imagenes': imagenes,
      'fecha_creacion': fechaCreacion?.toIso8601String(),
      'fecha_actualizacion': fechaActualizacion?.toIso8601String(),
      'ong_principal': ongPrincipal,
    };
  }

  bool get estaFinalizado {
    return estado == 'finalizado' || (fechaFin.isBefore(DateTime.now()));
  }

  bool get estaActivo {
    return activo && estado == 'activo' || estado == 'en_curso';
  }
}

File: app/Models/MegaEvento.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class MegaEvento extends Model
{
    use HasFactory;

    protected $table = 'mega_eventos';
    protected $primaryKey = 'mega_evento_id';
    public $timestamps = false; // Usamos fecha_creacion y fecha_actualizacion manualmente

    protected $fillable = [
        'titulo',
        'descripcion',
        'fecha_inicio',
        'fecha_fin',
        'ubicacion',
        'lat',
        'lng',
        'fecha_creacion',
        'fecha_actualizacion',
        'categoria',
        'estado',
        'ong_organizadora_principal',
        'capacidad_maxima',
        'es_publico',
        'activo',
        'imagenes'
    ];

    protected $casts = [
        'fecha_inicio' => 'datetime',
        'fecha_fin' => 'datetime',
        'fecha_creacion' => 'datetime',
        'fecha_actualizacion' => 'datetime',
        'activo' => 'boolean',
        'es_publico' => 'boolean',
        'imagenes' => 'array',
    ];

    /**
     * Accessor para obtener URLs completas de las imágenes
     * Se ejecuta después del cast 'array', así que $value ya es un array
     */
    public function getImagenesAttribute($value)
    {
        // Si $value es null o no es array, retornar array vacío
        if (!is_array($value)) {
            // Si es string, intentar decodificar JSON
            if (is_string($value)) {
                $decoded = json_decode($value, true);
                $value = is_array($decoded) ? $decoded : [];
            } else {
                $value = [];
            }
        }

        // Generar URLs completas para cada imagen
        return array_map(function($imagen) {
            if (empty($imagen) || !is_string($imagen)) {
                return null;
            }

            // Si ya es una URL completa, retornarla
            if (strpos($imagen, 'http://') === 0 || strpos($imagen, 'https://') === 0) {
                return $imagen;
            }

            // Si empieza con /storage/, agregar el dominio
            if (strpos($imagen, '/storage/') === 0) {
                return url($imagen);
            }

            // Si empieza con storage/, agregar /storage/
            if (strpos($imagen, 'storage/') === 0) {
                return url('/storage/' . $imagen);
            }

            // Por defecto, asumir que es relativa a storage
            return url('/storage/' . ltrim($imagen, '/'));
        }, array_filter($value, function($img) {
            return !empty($img) && is_string($img);
        }));
    }

    public function ongPrincipal()
    {
        return $this->belongsTo(Ong::class, 'ong_organizadora_principal', 'user_id');
    }

    /**
     * Boot del modelo para actualizar fecha_actualizacion automáticamente
     */
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($megaEvento) {
            if (empty($megaEvento->fecha_creacion)) {
                $megaEvento->fecha_creacion = now();
            }
            if (empty($megaEvento->fecha_actualizacion)) {
                $megaEvento->fecha_actualizacion = now();
            }
        });

        static::updating(function ($megaEvento) {
            $megaEvento->fecha_actualizacion = now();
        });
    }
}
